### PR TITLE
[indexer] Implemented CentersTable.

### PR DESCRIPTION
**Pull request by ygorshenin**
@mpimenov
@bykoianko 
PTAL

### CODE REVIEW

File: indexer/centers_table.cpp [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+  // CentersTable overrides:
+  bool Init() override
+  {
+    m_reader.Read(0, &m_header, sizeof(m_header));
+
+    if (!m_header.IsValid())
+      return false;
+
+    {
+      uint32_t idsSize = m_header.m_positionsOffset - m_header.m_identifiersOffset;
```
Comment (bykoianko) [inline indexer/centers_table.cpp]: const? Or probably it's worth adding an assert:

```
uint32_t const idsSize = m_header.m_positionsOffset - m_header.m_identifiersOffset;
ASSERT_LESS(m_header.m_identifiersOffset, m_header.m_positionsOffset, ());
```
?

File: indexer/centers_table.cpp [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+      vector<uint8_t> data(idsSize);
+      m_reader.Read(m_header.m_identifiersOffset, data.data(), data.size());
+      m_idsRegion = make_unique<CopiedMemoryRegion>(move(data));
+    }
+
+    {
+      uint32_t offsetsSize = m_header.m_deltasOffset - m_header.m_positionsOffset;
```
Comment (bykoianko) [inline indexer/centers_table.cpp]: const and assert as above?

File: indexer/centers_table.cpp [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+#include "coding/memory_region.hpp"
+#include "coding/reader.hpp"
+#include "coding/succinct_mapper.hpp"
+#include "coding/varint.hpp"
+#include "coding/writer.hpp"
+
+#include "base/assert.hpp"
+
+#include "std/unordered_map.hpp"
```
Comment (bykoianko) [inline indexer/centers_table.cpp]:
```
#include "std/unique_ptr.hpp"
```

File: indexer/centers_table.cpp [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+  vector<uint32_t> offsets;
+  vector<uint8_t> deltas;
+
+  {
+    MemWriter<vector<uint8_t>> writer(deltas);
+    for (size_t i = 0; i < m_centers.size(); i += CentersTableV0::kFrequency)
+    {
+      offsets.push_back(deltas.size());
+
+      uint64_t delta = EncodeDelta(m_centers[i], m_codingParams.GetBasePoint());
```
Comment (bykoianko) [inline indexer/centers_table.cpp]: const?

**Comment by bykoianko** [comment]
@ygorshenin I guess you've forgotten to add new coding/memory_region.hpp to this PR. Could you add it please?

File: indexer/centers_table.cpp (hunk 3 above)
Comment (ygorshenin) [inline indexer/centers_table.cpp]: It's included in the header.

File: indexer/centers_table.cpp (hunk 1 above)
Comment (ygorshenin) [inline indexer/centers_table.cpp]: No need in assert, as all these conditions are checked in m_header.IsValid()

File: indexer/centers_table.cpp (hunk 4 above)
Comment (ygorshenin) [inline indexer/centers_table.cpp]: No, it don't work this way, as it's modified in the loop.

File: indexer/centers_table.cpp (hunk 2 above)
Comment (ygorshenin) [inline indexer/centers_table.cpp]: const Done
No need in assert.

**Comment by ygorshenin** [comment]
memory_region.hpp is added.
PTAL

File: coding/memory_region.hpp [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+public:
+  virtual ~MemoryRegion() = default;
+
+  virtual uint64_t Size() const = 0;
+  virtual uint8_t const *ImmutableData() const = 0;
+};
+
+class MappedMemoryRegion : public MemoryRegion {
+public:
+  MappedMemoryRegion(FilesMappingContainer::Handle &&handle)
```
Comment (mpimenov) [inline coding/memory_region.hpp]: Add a space?

File: coding/memory_region.hpp [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+class CopiedMemoryRegion : public MemoryRegion {
+public:
+  CopiedMemoryRegion(vector<uint8_t> &&buffer) : m_buffer(move(buffer)) {}
+
+  // MemoryRegion overrides:
+  uint64_t Size() const override { return m_buffer.size(); }
+  uint8_t const *ImmutableData() const override { return m_buffer.data(); }
+
+  inline uint8_t *MutableData() { return m_buffer.data(); }
```
Comment (mpimenov) [inline coding/memory_region.hpp]: Вверх, до overrides?

File: coding/memory_region.hpp (hunk 5 above)
Comment (ygorshenin) [inline coding/memory_region.hpp]: Done

File: coding/memory_region.hpp (hunk 6 above)
Comment (ygorshenin) [inline coding/memory_region.hpp]: No, thanks.

**Comment by ygorshenin** [comment]
PTAL

File: indexer/centers_table.cpp [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+
+    auto & entry = m_cache[base];
+    if (entry.empty())
+    {
+      entry.resize(kFrequency);
+
+      uint32_t const start = m_offsets.select(base);
+      uint32_t const end = base + 1 < m_offsets.num_ones()
+                               ? m_offsets.select(base + 1)
+                               : m_header.m_endOffset - m_header.m_deltasOffset;
```
Comment (bykoianko) [inline indexer/centers_table.cpp]: clang-format?

**Comment by bykoianko** [comment]
LGTM with a small remark.

File: indexer/centers_table.cpp [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+  CentersTableV0(Reader & reader, serial::CodingParams const & codingParams)
+    : m_reader(reader), m_codingParams(codingParams)
+  {
+  }
+
+  // CentersTable overrides:
+  bool Get(uint32_t id, m2::PointD & center) override
+  {
+    if (id > m_ids.size() || !m_ids[id])
```
Comment (mpimenov) [inline indexer/centers_table.cpp]: What if ```id == m_ids.size()```?

File: indexer/centers_table.cpp [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+      m_reader.Read(m_header.m_deltasOffset + start, data.data(), data.size());
+
+      MemReader mreader(data.data(), data.size());
+      ReaderSource<MemReader> msource(mreader);
+
+      uint64_t delta = ReadVarUint<uint64_t>(msource);
+      entry[0] = DecodeDelta(delta, m_codingParams.GetBasePoint());
+
+      for (size_t i = 1; i < kFrequency && msource.Pos() < data.size(); ++i)
```
Comment (mpimenov) [inline indexer/centers_table.cpp]: ```msource.Size() > 0``` как вариант

File: indexer/centers_table.cpp (hunk 7 above)
Comment (ygorshenin) [inline indexer/centers_table.cpp]: Yes, it's clang-format.

File: indexer/centers_table.cpp (hunk 8 above)
Comment (ygorshenin) [inline indexer/centers_table.cpp]: Bad things happens. assert in debug, ud in release. Added regression test, fixed, thanks!

File: indexer/centers_table.cpp (hunk 9 above)
Comment (ygorshenin) [inline indexer/centers_table.cpp]: Good to know, thanks!

**Comment by ygorshenin** [comment]
PTAL

**Comment by ygorshenin** [comment]
@mpimenov PTAL

File: indexer/centers_table.cpp [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+#include "std/unordered_map.hpp"
+
+#include "3party/succinct/elias_fano.hpp"
+#include "3party/succinct/rs_bit_vector.hpp"
+
+namespace search
+{
+namespace
+{
+class CentersTableV0 : public CentersTable
```
Comment (mpimenov) [inline indexer/centers_table.cpp]: Please add a comment on the layout of this data structure.

File: coding/memory_region.hpp [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+  virtual ~MemoryRegion() = default;
+
+  virtual uint64_t Size() const = 0;
+  virtual uint8_t const * ImmutableData() const = 0;
+};
+
+class MappedMemoryRegion : public MemoryRegion
+{
+public:
+  MappedMemoryRegion(FilesMappingContainer::Handle && handle) : m_handle(move(handle)) {}
```
Comment (bykoianko) [inline coding/memory_region.hpp]: explicit?

Comment (bykoianko) [inline coding/memory_region.hpp]: For all constructor with one param.

File: coding/reader.hpp [hunk 12]
```diff
@@ -133,6 +133,52 @@ class ModelReaderPtr : public ReaderPtr<ModelReader>
...
+  }
+
+  // unique_ptr<Reader> SubReader(uint64_t size)
+  // {
+  //   uint64_t const pos = m_pos;
+  //   Skip(size);
+  //   return m_reader.SubReader(pos, size);
+  // }
+
+  // unique_ptr<Reader> SubReader() { return SubReader(Size()); }
```
Comment (bykoianko) [inline coding/reader.hpp]: Why not to remove this code?

File: indexer/centers_table.cpp [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+// 64 varuints, where each varuint represents an encoded delta of the
+// center from some prediction. For the first center in the block map
+// base point is used as a prediction, for all other centers in the
+// block previous center is used as a prediction. This allows to
+// decode block of 64 consecutive centers at one syscall, and cache
+// them in RAM.
+class CentersTableV0 : public CentersTable
+{
+public:
+  static const uint32_t kFrequency = 64;
```
Comment (bykoianko) [inline indexer/centers_table.cpp]: constexpr?

File: indexer/centers_table.cpp [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+    header.m_deltasOffset = m_writer.Pos();
+    m_writer.Write(deltas.data(), deltas.size());
+    header.m_endOffset = m_writer.Pos();
+  }
+
+  int64_t const endOffset = m_writer.Pos();
+
+  m_writer.Seek(startOffset);
+  m_writer.Write(&header, sizeof(header));
+  m_writer.Seek(endOffset);
```
Comment (bykoianko) [inline indexer/centers_table.cpp]: Do you thing it's worth returning write position to the end just before stop writing?

Comment (ygorshenin) [inline indexer/centers_table.cpp]: Yes, it's worth. There're no guarantees that the caller will not use writer after the call.

File: indexer/centers_table.cpp (hunk 13 above)
Comment (ygorshenin) [inline indexer/centers_table.cpp]: Do we really need constexprs here? What benefits they'll give us here?

File: coding/reader.hpp (hunk 12 above)
Comment (ygorshenin) [inline coding/reader.hpp]: Thanks for catching this, this code is not needed.

File: coding/memory_region.hpp (hunk 11 above)
Comment (ygorshenin) [inline coding/memory_region.hpp]: Done

File: indexer/centers_table.cpp (hunk 13 above)
Comment (bykoianko) [inline indexer/centers_table.cpp]: I do not think it's an important point but constexpr could be used in more cases. If you want leave it as a constant leave it as is.

File: indexer/centers_table.cpp (hunk 7 above)
Comment (bykoianko) [inline indexer/centers_table.cpp]: ok

File: indexer/centers_table.cpp (hunk 14 above)
Comment (bykoianko) [inline indexer/centers_table.cpp]: Ok

File: indexer/centers_table.cpp [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+
+#include "3party/succinct/elias_fano.hpp"
+#include "3party/succinct/rs_bit_vector.hpp"
+
+namespace search
+{
+namespace
+{
+template <typename TCont>
+void EndiannesAwareMap(bool endiannesMismatch, CopiedMemoryRegion & region, TCont & cont)
```
Comment (mpimenov) [inline indexer/centers_table.cpp]: s/s/ss/

Comment (ygorshenin) [inline indexer/centers_table.cpp]: Done

**Comment by ygorshenin** [comment]
@mpimenov @bykoianko PTAL

File: coding/reader.hpp [hunk 16]
```diff
@@ -133,6 +133,43 @@ class ModelReaderPtr : public ReaderPtr<ModelReader>
 };
 
 // Source that reads from a reader.
+class NonOwningReaderSource
+{
+public:
+  NonOwningReaderSource(Reader const & reader) : m_reader(reader), m_pos(0) {}
```
Comment (bykoianko) [inline coding/reader.hpp]: explicit?

File: indexer/centers_table.cpp [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+// 12                   end of section      4
+// 16                   identifiers table   positions offset - 16
+// positions offset     positions table     deltas offset - positions offset
+// deltas offset        deltas blocks       end of section - deltas offset
+//
+// All offsets are in little-endian format.
+//
+// Identifiers table is a bit-vector with rank-select table, where set
+// bits denote that centers for the corresponding features are in
+// table.  Identifiers table is stored in a native endianness.
```
Comment (mpimenov) [inline indexer/centers_table.cpp]: in native endianness?
in the native endianness?

File: indexer/centers_table.cpp [hunk 18]
```diff
@@ -0,0 +1,337 @@
...
+//
+// Positions table is a elias-fano table, where each entry corresponds
```
Comment (mpimenov) [inline indexer/centers_table.cpp]: an elias-fano, or even an Elias-Fano

File: indexer/centers_table.cpp [hunk 19]
```diff
@@ -0,0 +1,337 @@
...
+// to the start position of the centers block. Positions table is
+// stored in a native endianness.
+//
+// Deltas is a sequence of blocks, where each block is a sequence of
```
Comment (mpimenov) [inline indexer/centers_table.cpp]: where each block (probably with the exception of the last one) ?

File: indexer/centers_table.cpp (hunk 13 above)
Comment (ygorshenin) [inline indexer/centers_table.cpp]: There're no complex compile-time computations related to the constant. I'd like to leave it as is.

File: indexer/centers_table.cpp [hunk 20]
```diff
@@ -0,0 +1,337 @@
...
+// 64 varuints, where each varuint represents an encoded delta of the
```
Comment (mpimenov) [inline indexer/centers_table.cpp]: s/64/|kFrequency|/ как вариант

File: indexer/centers_table.cpp [hunk 21]
```diff
@@ -0,0 +1,337 @@
...
+  TCont c;
+  if (endiannesMismatch)
+  {
+    coding::ReverseMapVisitor visitor(region.MutableData());
+    c.map(visitor);
+  }
+  else
+  {
+    coding::MapVisitor visitor(region.ImmutableData());
+    c.map(visitor);
```
Comment (milchakov) [inline indexer/centers_table.cpp]: c.map(visitor); вынести из условных блоков

File: indexer/centers_table.cpp [hunk 22]
```diff
@@ -0,0 +1,337 @@
...
+  }
+
+  c.swap(cont);
+}
+
+// V0 of CentersTable.  Has the following format:
+//
+// File offset (bytes)  Field name          Field size (bytes)
```
Comment (mpimenov) [inline indexer/centers_table.cpp]: What about the first 4 bytes?

File: indexer/centers_table.cpp [hunk 23]
```diff
@@ -0,0 +1,337 @@
...
+  // CentersTable overrides:
+  bool Init() override
+  {
+    m_header.Read(m_reader);
+
+    if (!m_header.IsValid())
+      return false;
+
+    bool const isHostBigEndian = IsBigEndian();
+    bool const isDataBigEndian = m_header.m_base.m_endiannes == 1;
```
Comment (mpimenov) [inline indexer/centers_table.cpp]: s/m_endiannes/m_endianness/g

File: indexer/centers_table.cpp [hunk 24]
```diff
@@ -0,0 +1,337 @@
...
+  return true;
+}
+
+// CentersTable ------------------------------------------------------------------------------------
+unique_ptr<CentersTable> CentersTable::Load(Reader & reader,
+                                            serial::CodingParams const & codingParams)
+{
+  uint16_t const version = ReadPrimitiveFromPos<uint16_t>(reader, 0 /* pos */);
+  if (version != 0)
+    return unique_ptr<CentersTable>();
```
Comment (mpimenov) [inline indexer/centers_table.cpp]: Add a log?

File: coding/reader.hpp (hunk 16 above)
Comment (ygorshenin) [inline coding/reader.hpp]: Done

File: indexer/centers_table.cpp (hunk 17 above)
Comment (ygorshenin) [inline indexer/centers_table.cpp]: Done

File: indexer/centers_table.cpp (hunk 18 above)
Comment (ygorshenin) [inline indexer/centers_table.cpp]: Done

File: indexer/centers_table.cpp (hunk 19 above)
Comment (ygorshenin) [inline indexer/centers_table.cpp]: Done

File: indexer/centers_table.cpp (hunk 20 above)
Comment (ygorshenin) [inline indexer/centers_table.cpp]: Done

File: indexer/centers_table.cpp (hunk 22 above)
Comment (ygorshenin) [inline indexer/centers_table.cpp]: Basic header, documented in hpp file.

File: indexer/centers_table.cpp (hunk 23 above)
Comment (ygorshenin) [inline indexer/centers_table.cpp]: Done

File: indexer/centers_table.cpp (hunk 24 above)
Comment (ygorshenin) [inline indexer/centers_table.cpp]: Done

File: indexer/centers_table.hpp [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+//
+// *NOTE* This wrapper is abstract enough so feel free to change it,
+// but note that there should always be backward-compatibility.  Thus,
+// when adding new versions, never change data format of old versions.
+//
+// All centers tables are serialized in the following format:
+//
+// File offset (bytes)  Field name  Field size (bytes)
+// 0                    version     2
+// 2                    endiannes   2
```
Comment (mpimenov) [inline indexer/centers_table.hpp]: s/s/ss/

Comment (ygorshenin) [inline indexer/centers_table.hpp]: Done

**Comment by ygorshenin** [comment]
@mpimenov @bykoianko PTAL

**Comment by mpimenov** [comment]
LGTM

**Comment by bykoianko** [comment]
LGTM

**Comment by ygorshenin** [comment]
Thanks for the review!

**Pull request merged**